### PR TITLE
Path: Fix PointProjectionFailed

### PR DESCRIPTION
**Pull request by younghang**
fix the PointProjectionFailed exception when split edge in wrebaseWire().

---
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886">pipeline status</a> for feature branch [PR_6886](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886). Pipeline [539544670](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/539544670) was triggered at [7630ee0](https://github.com/FreeCAD/FreeCAD/pull/6886/commits/7630ee024be793496e0333e7d415022a027358bf). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
Is there a forum thread associated with this PR?

**Comment by luzpaz** [comment]
Looks like it's failing the CI, but I don't understand why.   
@younghang please run:  
`git commit --amend -m "Path: fix the PointProjectionFailed exception when split edge in wrebaseWire() -m "<reference a forum discussion here>"`

**Comment by younghang** [comment]
> Is there a forum thread associated with this PR?

No, I have searched this in the forum. I found the reason for the PointProjectionFailed, that is when
```C++
 gp_Pnt pprev(BRep_Tool::Pnt(xp.CurrentVertex()));
```
is called, there is a tiny difference between pprev and vertex around 0.0000001. which raise the PointProjectionFail exception in
 ```C++
 mkEdge1.Init(curve, myBestPt, pprev);
```
.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886">pipeline status</a> for feature branch [PR_6886](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886). Pipeline [539716433](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/539716433) was triggered at [07d7ced](https://github.com/FreeCAD/FreeCAD/pull/6886/commits/07d7ced8eca6c8715e2eb1d9bdf94b25b784f9fc). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Path/App/Area.cpp [hunk 1]
```diff
@@ -2615,7 +2615,7 @@ struct ShapeInfo{
                         GeomAPI_ProjectPointOnCurve gpp;
                         gpp.Init( myBestPt,curve);
                         gpp.Perform(myBestPt);
-                        myBestPt=c.NearestPoint();
+                        myBestPt= gpp.NearestPoint();
```
Comment (luzpaz) [inline src/Mod/Path/App/Area.cpp:2618]: Please precede the `=` with a whitespace as well.

**Comment by luzpaz** [comment]
CC @sliptonic please review when you get a chance..thanks!

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886">pipeline status</a> for feature branch [PR_6886](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886). Pipeline [540628320](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/540628320) was triggered at [5ae3a08](https://github.com/FreeCAD/FreeCAD/pull/6886/commits/5ae3a0866e187c27aa9cee33952f060d055b4cb8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Path/App/Area.cpp [hunk 2]
```diff
@@ -2615,7 +2615,7 @@ struct ShapeInfo{
                         GeomAPI_ProjectPointOnCurve gpp;
                         gpp.Init( myBestPt,curve);
```
Comment (luzpaz) [inline src/Mod/Path/App/Area.cpp:2616]: please fix whitespace here as well

**Comment by realthunder** [comment]
@younghang I suggest to try this projection on curve only if it fails to make either edges.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886">pipeline status</a> for feature branch [PR_6886](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6886). Pipeline [541638197](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/541638197) was triggered at [a21a5a2](https://github.com/FreeCAD/FreeCAD/pull/6886/commits/a21a5a20324dc9f3f477ef7289ab4ea23c9b0eb9). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Path/App/Area.cpp (hunk 2 above)
Comment (younghang) [reply src/Mod/Path/App/Area.cpp]: OKay, I have done it. It's my first time to commit a pull request on GitHub, some operations are not quiet familiar which may consume your time.

**Comment by younghang** [comment]
> curve only if it fails to make either edges.

@realthunder The split operation only executes once when the StartPoint is set, and I think the split point should on the edge though not all edges fail in current operation.

Another bug I found in this rebaseWire() is when the StartPoint is set on some edge by force, the closed wire will be broken some edges are missing.
I use 
```C++
       if (edge.Orientation() != myBestWire->wire.Orientation())
       {
            mkEdge1.Init(curve, myBestPt, pprev);
            mkEdge2.Init(curve, pt, myBestPt);
       }
```
instead of 
```C++
      if(reversed) {
           mkEdge1.Init(curve, myBestPt, pprev);
           mkEdge2.Init(curve, pt, myBestPt);
      }
```
and it seems to work in my project.

**Comment by luzpaz** [comment]
@younghang can you post to the [Path subforum](https://forum.freecadweb.org/viewforum.php?f=15) announcing this PR and x-post this URL back to this ticket so we can track it?)

**Comment by younghang** [comment]
> @younghang can you post to the [Path subforum](https://forum.freecadweb.org/viewforum.php?f=15) announcing this PR and x-post this URL back to this ticket so we can track it?)

@luzpaz Okay. I have post it here: https://forum.freecadweb.org/viewtopic.php?f=15&t=68956

**Comment by donovaly** [comment]
What is the status of this PR?

@sliptonic , can you please have a look?

**Comment by donovaly** [comment]
@younghang , many thanks for the fix!

In future you only need to check that it compiles also when precompiled headers.
I fixed this now by adding the missing include to Precompiled.h: https://github.com/FreeCAD/FreeCAD/commit/d13ecc1efa177be

**Pull request merged**